cloud/linode/route_controller.go

```diff
@@ -0,0 +1,237 @@
+package linode
+
+import (
+	"context"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/linode/linodego"
+	"golang.org/x/exp/slices"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cloudprovider "k8s.io/cloud-provider"
+	"k8s.io/klog/v2"
+)
+
+type routeCache struct {
+	sync.RWMutex
+	routes     map[int][]linodego.InstanceConfig
+	lastUpdate time.Time
+	ttl        time.Duration
+}
+
+func (rc *routeCache) refreshRoutes(ctx context.Context, client Client) error {
+	rc.Lock()
+	defer rc.Unlock()
+
+	if time.Since(rc.lastUpdate) < rc.ttl {
+		return nil
+	}
+
+	instances, err := client.ListInstances(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	rc.routes = make(map[int][]linodego.InstanceConfig, len(instances))
+	wg := sync.WaitGroup{}
+	for _, instance := range instances {
+		wg.Add(1)
+		go func(ctx context.Context, id int) {
+			configs, err := client.ListInstanceConfigs(ctx, id, &linodego.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed fetching instance configs for instance id %d. Error: %s", id, err.Error())
+			}
+			rc.routes[id] = configs
+			wg.Done()
+		}(ctx, instance.ID)
```

Comment:
I was thinking about this last night and I started to second guess about whether this would be ok as we know that the keys are unique, but long story short it is not safe. 


You can prove this by writing a simple test that uses go routines to concurrently write unique keys to a map and run the test with the race detector enabled. 

```
package linode

import (
	"testing"
)

func TestRefrshRoutes(t *testing.T) {
	m := map[int]rune{}

	for i := 0; i < 100; i++ {
		i := i
		go func(i int) {
			m[i] = rune(i)
		}(i)
	}
}

go test -race ./cloud/linode/route_controller_test.go
==================
WARNING: DATA RACE
Write at 0x00c000118570 by goroutine 8:
  runtime.mapaccess2_fast64()
      /opt/homebrew/Cellar/go/1.22.0/libexec/src/runtime/map_fast64.go:53 +0x1cc
  command-line-arguments.TestRefrshRoutes.func1()
      /Users/avest/Repos/linode-cloud-controller-manager/cloud/linode/route_controller_test.go:13 +0x3c
  command-line-arguments.TestRefrshRoutes.gowrap1()
      /Users/avest/Repos/linode-cloud-controller-manager/cloud/linode/route_controller_test.go:14 +0x44

Previous write at 0x00c000118570 by goroutine 10:
  runtime.mapaccess2_fast64()
      /opt/homebrew/Cellar/go/1.22.0/libexec/src/runtime/map_fast64.go:53 +0x1cc
  command-line-arguments.TestRefrshRoutes.func1()
      /Users/avest/Repos/linode-cloud-controller-manager/cloud/linode/route_controller_test.go:13 +0x3c
  command-line-arguments.TestRefrshRoutes.gowrap1()
      /Users/avest/Repos/linode-cloud-controller-manager/cloud/linode/route_controller_test.go:14 +0x44

Goroutine 8 (running) created at:
  command-line-arguments.TestRefrshRoutes()
      /Users/avest/Repos/linode-cloud-controller-manager/cloud/linode/route_controller_test.go:12 +0x48
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1742 +0x40

Goroutine 10 (finished) created at:
  command-line-arguments.TestRefrshRoutes()
      /Users/avest/Repos/linode-cloud-controller-manager/cloud/linode/route_controller_test.go:12 +0x48
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/Cellar/go/1.22.0/libexec/src/testing/testing.go:1742 +0x40
==================
fatal error: concurrent map writes
fatal error: concurrent map writes
fatal error: concurrent map writes
fatal error: concurrent map writes
```

Reply:
Uniqueness is not enough, because each write can trigger a resize and that is a no-no right there.